### PR TITLE
bad-arg tests for L3 functions

### PR DESCRIPTION
**Pull request by daineAMD**
Includes a couple of TODO's noting discrepancies. Will work on fixing TODOs for this release as well.

### CODE REVIEW

File: clients/include/blas3/testing_dgmm_batched.hpp [hunk 1]
```diff
@@ -38,6 +38,105 @@ inline void testname_dgmm_batched(const Arguments& arg, std::string& name)
...
+                                                   lda,
+                                                   dx.ptr_on_device(),
+                                                   incx,
+                                                   nullptr,
+                                                   ldc,
+                                                   batch_count),
+                              HIPBLAS_STATUS_INVALID_VALUE);
+    }
+
+    // If M == 0 || N == 0 || batch_count == 0, can have all nullptrs
```
Comment (NaveenElumalaiAMD) [inline clients/include/blas3/testing_dgmm_batched.hpp:131]: Just curious, I do not think so we have these bad arg tests in rocBLAS dgmm?

File: clients/include/blas3/testing_geam_batched.hpp [hunk 2]
```diff
@@ -50,6 +50,293 @@ inline void testname_geam_batched(const Arguments& arg, std::string& name)
...
+                                                     lda,
+                                                     zero,
+                                                     nullptr,
+                                                     ldb,
+                                                     dC.ptr_on_device(),
+                                                     ldc,
+                                                     batch_count));
+        }
+
+        // If M == 0 || N == 0, can have nullptrs
```
Comment (NaveenElumalaiAMD) [inline clients/include/blas3/testing_geam_batched.hpp]: If M == 0 || N == 0 || batch_count == 0

File: clients/include/blas3/testing_geam_strided_batched.hpp [hunk 3]
```diff
@@ -51,6 +51,344 @@ inline void testname_geam_strided_batched(const Arguments& arg, std::string& nam
...
+                                                            nullptr,
+                                                            ldb,
+                                                            strideB,
+                                                            dC,
+                                                            ldc,
+                                                            strideC,
+                                                            batch_count));
+        }
+
+        // If M == 0 || N == 0, can have nullptrs
```
Comment (NaveenElumalaiAMD) [inline clients/include/blas3/testing_geam_strided_batched.hpp]: batch_count == 0

File: clients/include/blas3/testing_hemm_batched.hpp [hunk 4]
```diff
@@ -47,6 +47,276 @@ inline void testname_hemm_batched(const Arguments& arg, std::string& name)
...
+                                                     lda,
+                                                     nullptr,
+                                                     ldb,
+                                                     one,
+                                                     nullptr,
+                                                     ldc,
+                                                     batch_count));
+        }
+
+        // If M == 0 || N == 0, can have nullptrs
```
Comment (NaveenElumalaiAMD) [inline clients/include/blas3/testing_hemm_batched.hpp]: batch_count ==0

File: clients/include/blas3/testing_her2k.hpp [hunk 5]
```diff
@@ -38,6 +38,147 @@ inline void testname_her2k(const Arguments& arg, std::string& name)
...
+                                             beta,
+                                             dC,
+                                             ldc),
+                              HIPBLAS_STATUS_INVALID_VALUE);
+        // EXPECT_HIPBLAS_STATUS(hipblasHer2kFn(handle, (hipblasFillMode_t)HIPBLAS_OP_N, transA, N, K, alpha, dA, lda, dB, ldb, beta, dC, ldc),
+        //                     HIPBLAS_STATUS_INVALID_ENUM);
+        // EXPECT_HIPBLAS_STATUS(
+        //     hipblasHer2kFn(
+        //         handle, uplo, HIPBLAS_OP_T, N, K, alpha, dA, lda, dB, ldb, beta, dC, ldc),
+        //     HIPBLAS_STATUS_INVALID_VALUE);
```
Comment (NaveenElumalaiAMD) [inline clients/include/blas3/testing_her2k.hpp:113]: why are these tests commented out?

File: clients/include/blas3/testing_her2k_batched.hpp [hunk 6]
```diff
@@ -47,6 +47,268 @@ inline void testname_her2k_batched(const Arguments& arg, std::string& name)
...
+                                                    lda,
+                                                    dB.ptr_on_device(),
+                                                    ldb,
+                                                    beta,
+                                                    dC.ptr_on_device(),
+                                                    ldc,
+                                                    batch_count),
+                              HIPBLAS_STATUS_INVALID_VALUE);
+        // EXPECT_HIPBLAS_STATUS(hipblasHer2kBatchedFn(handle, (hipblasFillMode_t)HIPBLAS_OP_N, transA, N, K, alpha, dA.ptr_on_device(), lda, dB.ptr_on_device(), ldb, beta, dC.ptr_on_device(), ldc, batch_count),
+        //                     HIPBLAS_STATUS_INVALID_ENUM);
```
Comment (NaveenElumalaiAMD) [inline clients/include/blas3/testing_her2k_batched.hpp:133]: same here

File: clients/include/blas3/testing_syrk_batched.hpp [hunk 7]
```diff
@@ -46,6 +46,182 @@ inline void testname_syrk_batched(const Arguments& arg, std::string& name)
...
+            // If k == 0 && beta == 1, A, C may be nullptr
+            CHECK_HIPBLAS_ERROR(hipblasSyrkBatchedFn(
+                handle, uplo, transA, N, 0, alpha, nullptr, lda, one, nullptr, ldc, batch_count));
+
+            // If alpha == 0 && beta == 1, A, C may be nullptr
+            CHECK_HIPBLAS_ERROR(hipblasSyrkBatchedFn(
+                handle, uplo, transA, N, K, zero, nullptr, lda, one, nullptr, ldc, batch_count));
+        }
+
+        // If N == 0, can have nullptrs
```
Comment (NaveenElumalaiAMD) [inline clients/include/blas3/testing_syrk_batched.hpp]: batch_count == 0

File: clients/include/blas3/testing_trmm_strided_batched.hpp [hunk 8]
```diff
@@ -184,148 +184,6 @@ void testing_trmm_strided_batched_bad_arg(const Arguments& arg)
...
-                                                          side,
-                                                          uplo,
-                                                          transA,
-                                                          diag,
-                                                          -1,
-                                                          N,
-                                                          alpha,
-                                                          dA,
-                                                          lda,
-                                                          strideA,
```
Comment (NaveenElumalaiAMD) [inline clients/include/blas3/testing_trmm_strided_batched.hpp:198]: strideA could be lda * K; when initialized in top

File: clients/gtest/blas3/dgmm_gtest.yaml [hunk 9]
```diff
@@ -45,4 +45,33 @@ Tests:
     stride_scale: [ 1.0, 2.5 ]
     api: [ FORTRAN, C ]
     backend_flags: AMD
+
+  - name: dgmm_bad_arg
+    category: pre_checkin
+    function:
+      - dgmm_bad_arg
```
Comment (NaveenElumalaiAMD) [inline clients/gtest/blas3/dgmm_gtest.yaml:52]: can functions `dgmm_batched_bad_arg` and `dgmm_strided_batched_bad_arg` added here. Similarly for all other Level 3 functions as well.

**Review by NaveenElumalaiAMD — APPROVED** [review]
LGTM! Minor comments raised.

File: clients/gtest/blas3/dgmm_gtest.yaml (hunk 9 above)
Comment (daineAMD) [reply clients/gtest/blas3/dgmm_gtest.yaml:52]: Hmm, I suppose they could. They were initially separated in the L1 PR before I added the `bad_arg_all` flag. I'll update in a commit today.

File: clients/include/blas3/testing_dgmm_batched.hpp (hunk 1 above)
Comment (daineAMD) [reply clients/include/blas3/testing_dgmm_batched.hpp:131]: Looks like no, but I do see the tests in the regular test suite/yaml file. Could handle this in yaml too but not sure if it makes a difference.

File: clients/include/blas3/testing_her2k.hpp (hunk 5 above)
Comment (daineAMD) [reply clients/include/blas3/testing_her2k.hpp:113]: See the comment in testing_herk.hpp. Just created LWPMLSE-516 to track this.

**Pull request merged**